docs/rfcs/read-only-replica.md

```diff
@@ -0,0 +1,60 @@
+# Read-only replicas
+
+We want to be able to spin-up read-only compute nodes.
+
+## Scope of this feature
+
+- We want to be able to spin-up multiple read-only replicas.
+- We should be able to spin-up RO node, if primary is inactive.
+
+- We don't want "snapshot" read-only computes at fixed LSN. Users can use branches for that.
+- We don't need replica promotion. For failower we rely on quick primary node restarts.
+
+- Implement it for v15 first.
+
+## Design
+
+### Replica creation
+
+1. At the moment of endpoint creation, user should be able to specify that they want to create a read-only endpoint.
+This setting will be passed via compute spec to `compute_ctl`. The replica will be created in the same region as primary (and all project's branches).
+
+Q: Should we allow to specify region for replica? It will affect latency and costs.
+
+2. `compute_ctl` will spin up a read-only replica.
+
+3. replica then connects to safekeepers and starts streaming WAL from them
+It would be good to connect to the safekeeper in the same AZ to reduce network latency.
+
+### Configuration & UI
+
+We take the same approach as with primary nodes and don't expose any configuration knobs to users.
+It may be useful to add a dashboard to show replica lag.
+
+### Replication
+
+- Replica receives WAL stream from safekeeper to keep up with primary.
```

Comment:
Also below,

>  - safekeepers should be able to stream WAL to multiple replicas and handle their feedback.

Similar things is [done](https://github.com/neondatabase/neon/tree/main/pageserver/src/tenant/timeline) by the pageserver now already: there's a lagging node detection and other heuristics.

A few related concerns:

* safekeepers share their state via broker, that uses `grpc` protocol. 

If I'm getting this right, the new functionality is supposed to appear in compute, so is it even feasible (possible?) to message over grpc from compute? 

* Is it so, that the current architecture still has pageserver as the only node connecting to safekeepers for data?
While pageserver has [some](https://github.com/neondatabase/neon/issues/2106) issues, it still aims at "get data form SKs ASAP and return a page for it" response handling.
Recent WAL (that's presumably understood under "to keep up with primary" activity) is stored in pageserver's memory and ready to serve page requests as soon as it's received from SK.

Since current proposal keeps the RO computes to pageserver, what are the benefits of making replicas to stream latest WAL from SKs directly?

Reply:
> Since current proposal keeps the RO computes to pageserver, what are the benefits of making replicas to stream latest WAL from SKs directly?

To clear/update internal caches (shared buffers and file cache). Read-only replica != proxy to pageserver

Reply:
> Replica receives WAL stream from safekeeper to keep up with primary.

To keep PG's SSI, it seems that when RW commits a transaction, RW has to wait until WAL is received and applied to all of replicas?

Reply:
Just waiting for applying WAL at all replicas is not enough to provide SSI. Some other mechanism (like commit serial numbers - CSN, or logical time) are need to enforce DSSI.

Standard postgres replication mechanism support async and sync mode. In the last case commit is blocked until all sync replicas confirm receiving the WAL. But not applying the WAL. So it is still possible with sync replication the changes visible at master are not yet visible at replica.